mp2-v1/src/values_extraction/planner.rs

```diff
@@ -0,0 +1,430 @@
+//! This code returns an [`UpdateTree`] used to plan how we prove a series of values was extracted from a Merkle Patricia Trie.
+#![allow(clippy::identity_op)]
+use alloy::{
+    network::Ethereum,
+    primitives::{keccak256, Address, B256},
+    providers::RootProvider,
+    transports::Transport,
+};
+use anyhow::Result;
+use mp2_common::eth::{node_type, EventLogInfo, NodeType, ReceiptQuery};
+use ryhope::storage::updatetree::{Next, UpdateTree};
+use std::future::Future;
+
+use std::collections::HashMap;
+
+use super::{generate_proof, CircuitInput, PublicParameters};
+/// Trait that is implemented for all data that we can provably extract.
+pub trait Extractable {
+    fn create_update_tree<T: Transport + Clone>(
+        &self,
+        contract: Address,
+        epoch: u64,
+        provider: &RootProvider<T, Ethereum>,
+    ) -> impl Future<Output = Result<UpdateTree<B256>>>;
+
+    fn prove_value_extraction<const MAX_COLUMNS: usize, T: Transport + Clone>(
+        &self,
+        contract: Address,
+        epoch: u64,
+        pp: &PublicParameters<512, MAX_COLUMNS>,
+        provider: &RootProvider<T, Ethereum>,
+    ) -> impl Future<Output = Result<Vec<u8>>>
+    where
+        [(); MAX_COLUMNS - 2]:,
+        [(); MAX_COLUMNS - 1]:,
+        [(); MAX_COLUMNS - 0]:;
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+struct ProofData {
+    node: Vec<u8>,
+    node_type: NodeType,
+    tx_index: Option<u64>,
+    proof: Option<Vec<u8>>,
+}
+
+impl ProofData {
+    pub fn new(node: Vec<u8>, node_type: NodeType, tx_index: Option<u64>) -> ProofData {
+        ProofData {
+            node,
+            node_type,
+            tx_index,
+            proof: None,
+        }
+    }
+}
+
+impl<const NO_TOPICS: usize, const MAX_DATA: usize> Extractable
+    for EventLogInfo<NO_TOPICS, MAX_DATA>
+where
+    [(); 7 - 2 - NO_TOPICS - MAX_DATA]:,
+{
+    async fn create_update_tree<T: Transport + Clone>(
+        &self,
+        contract: Address,
+        epoch: u64,
+        provider: &RootProvider<T, Ethereum>,
+    ) -> Result<UpdateTree<B256>> {
+        let query = ReceiptQuery::<NO_TOPICS, MAX_DATA> {
+            contract,
+            event: *self,
+        };
+
+        let proofs = query.query_receipt_proofs(provider, epoch.into()).await?;
+
+        // Convert the paths into their keys using keccak
+        let key_paths = proofs
+            .iter()
+            .map(|input| input.mpt_proof.iter().map(keccak256).collect::<Vec<B256>>())
+            .collect::<Vec<Vec<B256>>>();
+
+        // Now we make the UpdateTree
+        Ok(UpdateTree::<B256>::from_paths(key_paths, epoch as i64))
+    }
```

Comment:
Good API - but we need now to take into accounts retry.
I'm all ears for ideas. 
Maybe @hackaugusto has some as well (TLDR: we're doing more and more things internally so DQ has less to manage and we can test more - in this case, we'd be fetching directly the receipts and build all the inputs properly).

Retrying this method overall, with backoff, is the simplest for sure -  We'd be loosing time if there is another logic that fails inside (retry doesn't differentiate between specific errors, it just retries if it's an error or not). But maybe it's an acceptable loss for such a gain in API.